registry/remote/repository.go

```diff
@@ -193,28 +193,38 @@ func (r *Repository) FetchReference(ctx context.Context, reference string) (ocis
 // and returns the parsed reference. If the parsed reference does not share
 // the same base reference with the Repository r, ParseReference returns a
 // wrapped error ErrInvalidReference.
-func (r *Repository) ParseReference(reference string) (registry.Reference, error) {
-	ref, err := registry.ParseReference(reference)
-	if err != nil {
-		// reference is not a FQDN
-		if index := strings.IndexByte(reference, '@'); index != -1 {
-			// drop tag since the digest is present
-			reference = reference[index+1:]
-		}
+func (r *Repository) ParseReference(reference string) (ref registry.Reference, err error) {
+	ref, err = registry.ParseReference(reference)
+	if err != nil && len(reference) > 0 {
 		ref = registry.Reference{
 			Registry:   r.Reference.Registry,
 			Repository: r.Reference.Repository,
 			Reference:  reference,
 		}
-		if err = ref.ValidateReference(); err != nil {
-			return registry.Reference{}, err
+
+		// reference is not a FQDN
+		if index := strings.IndexByte(reference, '@'); index != -1 {
+			// drop tag since the digest is present
+			ref.Reference = reference[index+1:]
+			err = ref.ValidateReferenceAsDigest()
+		} else if index = strings.IndexByte(reference, ':'); index != -1 {
+			// no @ but :, so it can only be a digest
+			err = ref.ValidateReferenceAsDigest()
+		} else {
+			err = ref.ValidateReferenceAsTag()
+		}
```

Comment:
Simplifying the code since `ValidateReference()` will auto detect the reference type.
```suggestion
		// reference is not a FQDN
		if index := strings.IndexByte(reference, '@'); index != -1 {
			// drop tag since the digest is present
			ref.Reference = reference[index+1:]
			err = ref.ValidateReferenceAsDigest()
		} else {
			err = ref.ValidateReference()
		}
```

Reply:
See my earlier comment.  If `ValidateReference` is not given the boolean `isTag`, it will effectively suffer from amnesia.  When it doesn't know *how* the call was made (that there was no `@`), it cannot make the call that what it is about to assess *must not be a digest*.  It then looks at it, decides that it can't be a *tag*, but that it can be a *digest*.  Something akin to `TOCTOU` (https://cwe.mitre.org/data/definitions/367.html)

Reply:
Since `ValidateReference` is not invoked in `ParseReference` as mentioned in my earlier comment, there is no TOCTOU issue.

Reply:
With better understanding in `reference.go`, you may consider above suggestion again.